src/Elliptic/Systems/IrrotationalBns/Equations.cpp

```diff
@@ -0,0 +1,149 @@
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Index.hpp"
+#include "DataStructures/Tensor/EagerMath/DotProduct.hpp"
+#include "DataStructures/Tensor/Identity.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "Elliptic/Systems/IrrotationalBns/Geometry.hpp"
+#include "Elliptic/Systems/Poisson/Equations.hpp"
+#include "Utilities/GenerateInstantiations.hpp"
+#include "Utilities/MakeWithValue.hpp"
+
+namespace IrrotationalBns {
+
+void flat_potential_fluxes(
+    const gsl::not_null<tnsr::I<DataVector, 3>*> flux_for_potential,
+    const tnsr::II<DataVector, 3>& rotational_shift_stress,
+    const tnsr::i<DataVector, 3>& auxiliary_velocity) {
+  ::tenex::evaluate<ti::I>(
+      flux_for_potential,
+      -auxiliary_velocity(ti::j) * rotational_shift_stress(ti::I, ti::J));
+  for (size_t d = 0; d < 3; d++) {
+    flux_for_potential->get(d) += auxiliary_velocity.get(d);
+  }
+}
+
+void curved_potential_fluxes(
+    const gsl::not_null<tnsr::I<DataVector, 3>*> flux_for_potential,
+    const tnsr::II<DataVector, 3>& rotational_shift_stress,
+    const tnsr::II<DataVector, 3>& inverse_spatial_metric,
+    const tnsr::i<DataVector, 3>& auxiliary_velocity) {
+  ::tenex::evaluate<ti::I>(flux_for_potential,
+                           (inverse_spatial_metric(ti::I, ti::J) -
+                            rotational_shift_stress(ti::I, ti::J)) *
+                               auxiliary_velocity(ti::j));
+}
+
+void fluxes_on_face(
+    gsl::not_null<tnsr::I<DataVector, 3>*> face_flux_for_potential,
+    const tnsr::i<DataVector, 3>& face_normal,
+    const tnsr::I<DataVector, 3>& face_normal_vector,
+    const tnsr::II<DataVector, 3>& rotational_shift_stress,
+    const Scalar<DataVector>& velocity_potential) {
+  ::tenex::evaluate<ti::I>(
+      face_flux_for_potential,
+      velocity_potential() *
+          (face_normal_vector(ti::I) -
+           rotational_shift_stress(ti::I, ti::J) * face_normal(ti::j)));
+}
+
+void add_flat_potential_sources(
+    const gsl::not_null<Scalar<DataVector>*> source_for_potential,
+    const tnsr::i<DataVector, 3>& log_deriv_lapse_over_specific_enthalpy,
+    const tnsr::I<DataVector, 3>& flux_for_potential) {
+  for (size_t i = 0; i < 3; i++) {
+    source_for_potential->get() -=
+        flux_for_potential.get(i) *
+        log_deriv_lapse_over_specific_enthalpy.get(i);
+  }
+}
+
+void add_curved_potential_sources(
+    const gsl::not_null<Scalar<DataVector>*> source_for_potential,
+    const tnsr::i<DataVector, 3>& log_deriv_of_lapse_over_specific_enthalpy,
+    const tnsr::i<DataVector, 3>& christoffel_contracted,
+    const tnsr::I<DataVector, 3>& flux_for_potential) {
+  // The flux for the potential is the auxiliary velocity
+  // (when the index is in the right place)
+  ::tenex::update(source_for_potential,
+                  (*source_for_potential)() +
+                      flux_for_potential(ti::I) *
+                          (log_deriv_of_lapse_over_specific_enthalpy(ti::i) -
```

Comment:
Does this include the factor of `\rho` as well in `D_i \ln(\alpha\rho / h)`?

Also I think the Christoffels should be added, not subtracted? The equation is `-\partial_i F^i + S = f` please check.

Reply:
Yes, it does include the factor of `\rho`, I'll change the variable name to make that obvious but maybe we can condense it in the future.

Hm, I would think the Christoffel should be be subtracted,  we have the equation which looks like `-D_i F^i  + S_flat = f ` and then we expand the term `D_i F^i  = \partial_i F^i + Gamma^i_{ij}F^j` so `-Gamma^i_{ij} F^j` gets added to the sources

Reply:
Yes you're right it should be subtracted, but the other term as well if I did the math correctly